src/main/scala/org/broadinstitute/dsde/firecloud/service/LibraryService.scala

```diff
@@ -142,19 +148,38 @@ class LibraryService (protected val argUserInfo: UserInfo,
     }
   }
 
+  private def isInvalidWorkspace(workspace: Workspace): (Boolean, Option[String]) = {
+    val invalidMetadata = workspace.attributes.
+      get(AttributeName(AttributeName.libraryNamespace, "invalidDataset"))
+    if (!invalidMetadata.isDefined || invalidMetadata == Some(AttributeBoolean(true))) {
+      // recheck for required attributes
+      // if the invalidMetadata was not set and the workspace is invalid, should we persist it here?
+      isInvalid(workspace.attributes.toJson.compactPrint)
+    }
+    else (false, None)
+  }
+
   // should only be used to change published state
   def setWorkspaceIsPublished(ns: String, name: String, value: Boolean): Future[PerRequestMessage] = {
     rawlsDAO.getWorkspace(ns, name) flatMap { workspaceResponse =>
       val pub = isPublished(workspaceResponse)
       if (pub == value)
         Future(RequestComplete(NoContent))
       else {
-        rawlsDAO.updateLibraryAttributes(ns, name, updatePublishAttribute(value)) map { ws =>
-          if (value)
-            publishDocument(ws)
-          else
-            removeDocument(ws)
-          RequestComplete(ws)
+        val (invalid, errorMessage) = if (value)
+          // only need to check for valid attributes if we are actually publishing
+          isInvalidWorkspace(workspaceResponse.workspace)
+        else (false, None)
+        if (invalid) {
+          Future(RequestCompleteWithErrorReport(BadRequest, errorMessage.getOrElse(BadRequest.defaultMessage)))
+        } else {
+          rawlsDAO.updateLibraryAttributes(ns, name, updatePublishAttribute(value)) map { ws =>
+            if (value)
+              publishDocument(ws)
+            else
+              removeDocument(ws)
+            RequestComplete(ws)
+          }
```

Comment:
Stylistic nitpick. There are four if conditions, three of them nested in this method. I would try to make this more scala-idiomatic. Haven't tested this, but something like this reads more cleanly to me. Definitely open to criticism on this.
```
      isPublished(workspaceResponse) match {
        case x if x == value => Future(RequestComplete(NoContent))
        case _ =>
          // only need to check for valid attributes if we are actually publishing
          val (invalid, errorMessage) = value match {
            case x if x => isInvalidWorkspace(workspaceResponse.workspace)
            case _ => (false, None)
          }
          invalid match {
            case x if x => Future(RequestCompleteWithErrorReport(BadRequest, errorMessage.getOrElse(BadRequest.defaultMessage)))
            case _ =>
              rawlsDAO.updateLibraryAttributes(ns, name, updatePublishAttribute(value)) map { ws =>
                if (value)
                  publishDocument(ws)
                else
                  removeDocument(ws)
                RequestComplete(ws)
              }
          }
      }
```